src/d2lfetch-dedupe.js

```diff
@@ -1,51 +1,143 @@
+				: null;
+
+			const inflightRequest = new Request(request, {
+				signal: (abortController || {}).signal
+			});
 
-		if (!next) {
-			return Promise.resolve(request);
+			this._inflightRequests[key] = {
+				inflightRequest,
+				abortController,
+				dedupedRequests: {}
+			};
+
+			newRequest = true;
 		}
 
-		const result = next(request);
-		if (result && result instanceof Promise) {
-			this._inflightRequests[key] = { count: 1 };
-			this._inflightRequests[key].action = result
-				.then(function(response) {
-					const usedMultiple = this._inflightRequests[key].count !== 1;
-					delete this._inflightRequests[key];
-					return this._clone(response, usedMultiple);
-				}.bind(this))
-				.catch(function(err) {
-					delete this._inflightRequests[key];
-					return Promise.reject(err);
-				}.bind(this));
-
-			return this._inflightRequests[key].action;
+		const dedupedRequest = new Promise((resolve, reject) => {
+			const inflightRequestEntry = this._inflightRequests[key];
+
+			inflightRequestEntry.dedupedRequests[reqId] = {
+				resolvers: { resolve, reject }
+			};
+
+			if (request.signal && typeof request.signal.addEventListener === 'function') {
+				/**
+				 * @param {AbortSignal} signal
+				 * @param {string} key
+				 * @param {number} reqId
+				 */
+				const addAbortEventListener = (signal, key, reqId) => {
+					const handler = () => {
+						signal.removeEventListener('abort', handler);
+
+						const inflightRequestEntry = this._inflightRequests[key];
+
+						if (!inflightRequestEntry) {
+							return;
+						}
+
+						if (Object.keys(inflightRequestEntry.dedupedRequests).length === 1) {
+							const abortController = inflightRequestEntry.abortController;
+
+							if (abortController && typeof abortController.abort === 'function') {
+								abortController.abort();
+							}
+						}
+
+						delete inflightRequestEntry.dedupedRequests[reqId];
+						reject(new DOMException('Request was aborted.', 'AbortError'));
+					};
+
+					inflightRequestEntry.dedupedRequests[reqId].removeAbortListener = () => signal.removeEventListener('abort', handler);
+					signal.addEventListener('abort', handler);
+				};
+
+				addAbortEventListener(request.signal, key, reqId);
+			}
+		});
+
+		if (!newRequest) {
+			return dedupedRequest;
 		}
 
-		return result;
+		next(this._inflightRequests[key].inflightRequest)
+			.then((response) => {
+				const dedupedRequests = Object.values(this._inflightRequests[key].dedupedRequests);
+				const res = dedupedRequests.length > 1
+					? this._clone(response)
+					: response;
+
+				for (const dedupedRequest of dedupedRequests) {
+					if (typeof dedupedRequest.removeAbortListener === 'function') {
+						dedupedRequest.removeAbortListener();
+					}
+					dedupedRequest.resolvers.resolve(res);
+				}
+
+				delete this._inflightRequests[key];
+			})
+			.catch((err) => {
+				for (const dedupedRequest of Object.values(this._inflightRequests[key].dedupedRequests)) {
```

Comment:
Is there a chance that this catch block could run before the `abort` event gets handled? My understanding of Javascript would be no, but I'm not too sure.

Thoughts?

Reply:
This should be fine, but might be good to move `delete inflightRequestEntry.dedupedRequests[reqId];` further up in the `abort` handler, so you don't have to think about ticks to determine this is okay.

Reply:
I reasoned it wouldn't be an issue when writing it, but I agree with both, removing the hint of possibility is definitely preferable. Thanks!